kmcd-driver.md

```diff
@@ -124,19 +124,38 @@ On `exception`, the entire current call is discarded to trigger state roll-back
     syntax CallFrame ::= frame(prevSender: Address, prevEvents: List, continuation: K)
  // ----------------------------------------------------------------------------------
 
-    syntax AdminStep ::= "call" MCDStep
- // -----------------------------------
-    rule <k> call MCD:MCDStep ~> CONT => MCD </k>
+    syntax AdminStep ::= "call"        MCDStep
+                       | "checkauth"   MCDStep
+                       | "checklock"   MCDStep
+                       | "checkunlock" MCDStep
+                       | "makecall"    MCDStep
+ // ------------------------------------------
+
+    syntax MCDStep  ::= LockStep
+    syntax LockStep ::= "lock"   MCDStep
+                      | "unlock" MCDStep
+ // ------------------------------------
+    rule <k> call MCD:MCDStep => checkauth MCD ~> checklock MCD ~> makecall MCD ~> checkunlock MCD ...  </k>
+
+    rule <k> makecall MCD:MCDStep ~> CONT => MCD </k>
          <msg-sender> MSGSENDER => THIS </msg-sender>
          <this> THIS => contract(MCD) </this>
          <call-stack> .List => ListItem(frame(MSGSENDER, EVENTS, CONT)) ... </call-stack>
          <frame-events> EVENTS => ListItem(LogNote(MSGSENDER, MCD)) </frame-events>
-      requires isAuthStep(MCD) impliesBool isAuthorized(THIS, contract(MCD))
 
-    rule <k> call MCD => exception MCD ... </k> [owise]
+    rule <k> makecall MCD => exception MCD ... </k> [owise]
 
     rule <k> V:Value => . ... </k> <return-value> _ => V </return-value>
 
+    rule <k> checkauth MCD:AuthStep => .             ... </k> <this> THIS </this> requires isAuthorized(THIS, contract(MCD))
+    rule <k> checkauth MCD          => .             ... </k>                     requires notBool isAuthStep(MCD)
+    rule <k> checkauth MCD          => exception MCD ... </k>                     [owise]
+
+    rule <k> checklock MCD:LockStep => lock MCD ... </k>
+    rule <k> checklock MCD          => .        ... </k> requires notBool isLockStep(MCD)
+
+    rule <k> checkunlock MCD:LockStep => unlock MCD ... </k>
+    rule <k> checkunlock MCD          => .          ... </k> requires notBool isLockStep(MCD)
     rule <k> . => CONT </k>
          <msg-sender> MSGSENDER => PREVSENDER </msg-sender>
```

Comment:
Re-organize these rules a bit please. The one below `<k> . => CONT </k>` is for returning from a `makecall`, so should be grouped with that rule above.

Also, you should have generic rules which say:

```
rule <k> lock   MCD => exception MCD ... </k> [owise]
rule <k> unlock MCD => exception MCD ... </k> [owise]
```

The second one is not necessary, and the solidity code does not enforce it, but you can include it for completeness.

It's up to each individual declaration of a `LockStep` to make sure that the positive case (non-exception case) is declared correctly.

In addition, the steps `"lock"` and `"unlock"` should be `AdminStep`, not `LockStep`. The `LockStep` sort is to be declared on each individual function which needs that modifier.

Reply:
Also, please add the sort `LockAuthStep`, and complete the sort heirarchy.